crates/pipeline-manager/src/api/main.rs

```diff
@@ -59,14 +59,116 @@ out of the SQL program, but also has accompanying metadata and configuration par
+* Pipeline deployment start is only able to proceed to provisioning once the program is successfully
+  compiled.
 * A pipeline cannot be updated while it is deployed.
 
 ## Concurrency
 
 Each pipeline has a version, which is incremented each time its core fields are updated.
 The version is monotonically increasing. There is additionally a program version which covers
-only the program-related core fields, and is used by the compiler to discern when to recompile."
+only the program-related core fields, and is used by the compiler to discern when to recompile.
+
+## Client request handling
+
+### Request outcome expectations
+
+The outcome of a request is that it either fails (e.g., DNS lookup failed) without any response
+(no status code nor body), or it succeeds and gets back a response status code and body.
+
+In case of a response, usually it is the Feldera endpoint that generated it:
+- If it is success (2xx), it will return whichever body belongs to the success response.
+- Otherwise, if it is an error (4xx, 5xx), it will return a Feldera error response JSON body
+  which will have an application-level `error_code`.
+
+However, there are two notable exceptions when the response is not generated by the Feldera
+endpoint:
+- If the HTTP server, to which the endpoint belongs, encountered an issue, it might return
+  4xx (e.g., for an unknown endpoint) or 5xx error codes by itself (e.g., when it is initializing).
+- If the Feldera API server is behind a (reverse) proxy, the proxy can return error codes by itself,
+  for example BAD GATEWAY (502) or GATEWAY TIMEOUT (504).
+
+As such, it is not guaranteed that the (4xx, 5xx) will have a Feldera error response JSON body
+in these latter cases.
+
+### Error handling and retrying
+
+The error type returned by the client should distinguish between the error responses generated
+by Feldera endpoints themselves (which have a Feldera error response body) and those that are
+generated by other sources.
+
+In order for a client operation (e.g., `pipeline.resume()`) to be robust (i.e., not fail due to
+a single HTTP request not succeeding) the client should use a retry mechanism if the operation
+is idempotent. The retry mechanism must however have a time limit, after which it times out.
+This guarantees that the client operation is eventually responsive, which enables the script
+it is a part of to not hang indefinitely on Feldera operations and instead be able to decide
+by itself whether and how to proceed. If no response is returned, the mechanism should generally
+retry. When a response is returned, the decision whether to retry can generally depend on the status
+code: especially the status codes 502, 503 and 504 should be considered as transient errors.
+Finer grained retry decisions should be made by taking into account the application-level
+`error_code` if the response body was indeed a Feldera error response body.
```

Comment:
I wonder if the backend should include a `retry` field in the error message, based on which all the different clients can decide to retry.